clay/random.clj

```diff
@@ -1,11 +1,2259 @@
+                      [:integer-discrete-distribution (r/distribution-parameters (r/distribution :integer-discrete-distribution) true) false]
+                      [:categorical-distribution (r/distribution-parameters (r/distribution :categorical-distribution) true) false]
+                      [:enumerated-real (r/distribution-parameters (r/distribution :enumerated-real) true) false]
+                      [:enumerated-int (r/distribution-parameters (r/distribution :enumerated-int) true) false]])})
+
+
+;; #### Continuous
+
+;; Continous distribution build from data is based on KDE (Kernel Density Estimation) and PDF integration for CDF and iCDF. Mean and variance are calculated from samples.
+
+;; `:continous-distribution` and `:kde` are two names for the same distribution
+
+;; * Name: `:continuous-distribution` or `:kde`
+;; * Default parameters:
+;;    * `:data`, samples, sequence of numbers
+;;    * `:kde`, density estimation kernel: `:epenechnikov`
+;;    * `:bandwidth`, KDE bandwidth, smoothing parameter: `nil` (auto)
+;;    * `:steps`, number of steps for PDF integration: 5000
+;;    * `:min-iterations`, number of PDF integrator iterations: `nil` (default)
+;;    * `:interpolator`, CDF/iCDF interpolator for PDF integration: `nil` (default, `:linear`)
+;; * [wiki](https://en.wikipedia.org/wiki/Kernel_density_estimation), [pdf integration](#pdf-integration)
+
+(def random-data (repeatedly 1000 (fn [] (+ (* (r/drand -2 2) (r/drand -2 2))
+                                            (m/sqrt (* (r/drand) (r/drand)))))))
+
+(def kde-distr (r/distribution :continuous-distribution {:data random-data}))
+
+(let [build-distr (fn [pars]
+                    (r/distribution :continuous-distribution (assoc pars :data random-data)))]
+  (utls/dgraph-table
+   [["default", "bandwidth=1.0"]
+    [(mapv gg/->image (gg/dgraph kde-distr {:pdf [-5 5]}))
+     (mapv gg/->image (gg/dgraph (build-distr {:bandwidth 1.0}) {:pdf [-5 5]}))]
+    ["gaussian kernel" "triangular kernel, bandwidth=0.1"]
+    [(mapv gg/->image (gg/dgraph (build-distr {:kde :gaussian}) {:pdf [-5 5]}))
+     (mapv gg/->image (gg/dgraph (build-distr {:kde :triangular :bandwidth 0.1}) {:pdf [-5 5]}))]]))
+
+(utls/examples-note
+ (r/sample kde-distr)
+ (r/sample kde-distr)
+ (r/pdf kde-distr 0.0)
+ (r/cdf kde-distr 0.0)
+ (map (partial r/icdf kde-distr) [0.0001 0.5 0.9999])
+ (r/lower-bound kde-distr)
+ (r/upper-bound kde-distr)
+ (r/mean kde-distr)
+ (r/variance kde-distr))
+
+;; ##### Kernels
+
+;; Distributions for various kernels:
+;; * `:data`: `[-2 -2 -2 -1 0 1 2 -1 0 1 2 0 1 2 1 2 2]`
+;; * `:steps`: `100`
+;; * `:bandwidth`: auto
+
+;; `fastmath.kernel/kernel-list` contains three types of kernels: RBF, KDE and what we can call "vector kernels" which includes Marcer, positive definite, and similar.
+
+(kind/table
+ {:column-names ["Type" "Kernels"]
+  :row-vectors (->> k/kernel-list
```

Comment:
List only `:kde` key, The rest is irrelevant here.

Reply:
Yes, now it's only `:kde` kernels.